File: reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode newHead;
    ListNode ret=null;
    public ListNode reverseList(ListNode head) {
        newHead=null;
        dfs(head);
        return ret;
    }
    
    public void dfs(ListNode head){
        if(head==null)
            return;
        dfs(head.next);
        if(newHead==null){
            newHead=head;
            ret=head;
        }else{
            newHead.next=head;
            newHead=head;
        }
        newHead.next=null;
    }
}

File: reverse-linked-list/reverse-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* reverseList(struct ListNode* head){
    if(head==NULL)
        return NULL;
    struct ListNode* prev=head;
    head=head->next;
    prev->next=NULL;
    
    while(head!=NULL){
        struct ListNode* temp=head->next;  
        head->next=prev;
        prev=head;
        head=temp;
    }
    return prev;
}